lib/Sema/CSDiag.cpp

```diff
@@ -6508,7 +6510,100 @@ bool FailureDiagnosis::visitClosureExpr(ClosureExpr *CE) {
         }
         return true;
       }
-      
+
+      if (inferredArgCount == 1 && actualArgCount > 1) {
+        // Let's see if inferred argument is actually a tuple inside of Paren.
+        if (auto *argTupleTy = inferredArgType->getAs<TupleType>()) {
+          // Looks like the number of closure parameters matches number
+          // of inferred arguments, which means we can we can emit an
+          // error about an attempt to make use of tuple splat or tuple
+          // destructuring and provide a proper fix-it.
+          if (argTupleTy->getNumElements() == actualArgCount) {
+            // In case of implicit parameters e.g. $0, $1 we
+            // can't really provide good fix-it because
+            // structure of parameter type itself is unclear.
+            for (auto *param : params->getArray()) {
+              if (param->isImplicit()) {
+                diagnose(params->getStartLoc(),
+                         diag::closure_tuple_parameter_destructuring_implicit,
+                         argTupleTy);
+                return true;
+              }
+            }
+
+            auto diag = diagnose(params->getStartLoc(),
+                                 diag::closure_tuple_parameter_destructuring,
+                                 argTupleTy);
+            Type actualArgType;
+            if (auto *actualFnType = CE->getType()->getAs<AnyFunctionType>())
+              actualArgType = actualFnType->getInput();
+
+            auto &sourceMgr = CS->getASTContext().SourceMgr;
+
+            SmallString<64> fixIt;
+            llvm::raw_svector_ostream OS(fixIt);
+
+            // If this is multi-line closure we'd have to insert new lines
+            // in the suggested 'let' to keep the structure of the code intact,
+            // otherwise just use ';' to keep everything on the same line.
+            bool isMultiLineClosure = false;
+            StringRef indent = "";
+            if (auto *body = CE->getBody()) {
+              auto inLine = sourceMgr.getLineNumber(CE->getInLoc());
+              auto bodyLine = sourceMgr.getLineNumber(body->getStartLoc());
+              isMultiLineClosure = bodyLine > inLine;
+              indent =
+                  Lexer::getIndentationForLine(sourceMgr, body->getStartLoc());
+              ;
+            }
+
+            SmallString<16> parameter;
+            llvm::raw_svector_ostream parameterOS(parameter);
+
+            parameterOS << "(";
+            interleave(params->getArray(),
+                       [&](const ParamDecl *param) {
+                         parameterOS << param->getNameStr();
+                       },
+                       [&] { parameterOS << ", "; });
+            parameterOS << ")";
+
+            // If at least one of the parameters has explicitly
+            // declared type let's turn fix-it into N 'let' expressions
+            // to be able to specify it.
```

Comment:
I don't believe it is - because in situation like map { $0 + $1 } we are not going to provide fix-it just say that this is tuple restructuring with implicit parameters and it's not supported...